src/TimingData.h

```diff
@@ -193,50 +193,313 @@ class TimingData
 		return GetSegmentAtRow( BeatToNoteRow(fBeat), tst );
 	}
 
-	#define DefineSegmentWithName(Seg, SegName, SegType) \
-		const Seg* Get##Seg##AtRow( int iNoteRow ) const \
-		{ \
-			const TimingSegment *t = GetSegmentAtRow( iNoteRow, SegType ); \
-			return To##SegName( t ); \
-		} \
-		Seg* Get##Seg##AtRow( int iNoteRow ) \
-		{ \
-			return const_cast<Seg*> (((const TimingData*)this)->Get##Seg##AtRow(iNoteRow) ); \
-		} \
-		const Seg* Get##Seg##AtBeat( float fBeat ) const \
-		{ \
-			return Get##Seg##AtRow( BeatToNoteRow(fBeat) ); \
-		} \
-		Seg* Get##Seg##AtBeat( float fBeat ) \
-		{ \
-			return const_cast<Seg*> (((const TimingData*)this)->Get##Seg##AtBeat(fBeat) ); \
-		} \
-		void AddSegment( const Seg &seg ) \
-		{ \
-			AddSegment( &seg ); \
-		}
+		// BPM / BPM
+
+	const BPMSegment* GetBPMSegmentAtRow(int iNoteRow) const
+	{
+		const TimingSegment* t = GetSegmentAtRow(iNoteRow, SEGMENT_BPM);
+		return ToBPM(t);
+	}
+	
+	BPMSegment* GetBPMSegmentAtRow(int iNoteRow)
+	{
+		return const_cast<BPMSegment*> (((const TimingData*)this)->GetBPMSegmentAtRow(iNoteRow));
+	}
+	
+	const BPMSegment* GetBPMSegmentAtBeat(float fBeat) const
+	{
+		return GetBPMSegmentAtRow(BeatToNoteRow(fBeat));
+	}
+	
+	BPMSegment* GetBPMSegmentAtBeat(float fBeat)
+	{
+		return const_cast<BPMSegment*> (((const TimingData*)this)->GetBPMSegmentAtBeat(fBeat));
+	}
+	
+	void AddSegment(const BPMSegment& seg)
+	{
+		AddSegment(&seg);
+	}
+
+		// Stop / STOP
+
+	const StopSegment* GetStopSegmentAtRow(int iNoteRow) const
+	{
+		const TimingSegment* t = GetSegmentAtRow(iNoteRow, SEGMENT_STOP);
+		return ToStop(t);
+	}
+	
+	StopSegment* GetStopSegmentAtRow(int iNoteRow)
+	{
+		return const_cast<StopSegment*>(((const TimingData*)this)->GetStopSegmentAtRow(iNoteRow));
+	}
+	
+	const StopSegment* GetStopSegmentAtBeat(float fBeat) const
+	{
+		return GetStopSegmentAtRow(BeatToNoteRow(fBeat));
+	}
+	
+	StopSegment* GetStopSegmentAtBeat(float fBeat)
+	{
+		return const_cast<StopSegment*>(((const TimingData*)this)->GetStopSegmentAtBeat(fBeat));
+	}
+	
+	void AddSegment(const StopSegment& seg)
+	{
+		AddSegment(&seg);
+	}
+
+		// Delay / DELAY
+
+	const DelaySegment* GetDelaySegmentAtRow(int iNoteRow) const
+	{
+		const TimingSegment* t = GetSegmentAtRow(iNoteRow, SEGMENT_DELAY);
+		return ToDelay(t);
+	}
+	
+	DelaySegment* GetDelaySegmentAtRow(int iNoteRow)
+	{
+		return const_cast<DelaySegment*>(((const TimingData*)this)->GetDelaySegmentAtRow(iNoteRow));
+	}
+	
+	const DelaySegment* GetDelaySegmentAtBeat(float fBeat) const
+	{
+		return GetDelaySegmentAtRow(BeatToNoteRow(fBeat));
+	}
+	
+	DelaySegment* GetDelaySegmentAtBeat(float fBeat)
+	{
+		return const_cast<DelaySegment*>(((const TimingData*)this)->GetDelaySegmentAtBeat(fBeat));
+	}
+	
+	void AddSegment(const DelaySegment& seg)
+	{
+		AddSegment(&seg);
```

Comment:
Can't understand this code. Is it a endless recursion? Was macro expected to be inside of  NotesLoader class?

Reply:
It's not endless recursion because it's calling the pointer variant `AddSegment(const TimingSegment *seg)` although I admit it is a bit confusing. Might be worth a name change.

Reply:
This is not too uncommon in the code base unfortunately. There are a lot of variants of functions that share the same name but differ by what arguments they take.  It could definitely be improved for readability.  

With that said, this PR is just the expanded macro and nothing else. If @teejusb wants them renamed in this PR I'll do that though.